Inline child array and map data slab into parent slab

fxamacker opened the pull request:
Updates epics #292 https://github.com/onflow/flow-go/issues/1744
Updates #296 #297 https://github.com/onflow/cadence/issues/1854

## Problem

Memory use on execution nodes, register count, and growth rate can be improved.

Currently, every array or map are stored in its own slab and parent slab refers to child array or map by SlabID.

The current approach can lead to many small slabs, especially for Cadence data structures with multiple nested levels.

## Changes

Reduce register count and optimize encoding size.

This commit inlines child array/map in parent slab when:
- child array/map fits in one slab (root slab is data slab)
- encoded size of inlined child array/map is less than the max inline size limit enforced by parent

This commit optimizes encoding size by:
- reusing inlined array types
- reusing seed, digests, and field names of inlined composite types

Also update debugging code to handle inlined array/map element.

## TODO

### In this PR

- [x] add more tests
- [x] cleanup some of the tests

### Atree Inlining Part 2

Open new PR to add support for inlinability of child values returned by iterator (requires PR #329)

### Cadence Integration

Open new PR in onflow/cadence to integrate with Cadence.

______

<!-- Complete: -->

- [x] Targeted PR against `main` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/atree/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

turbolent commented:
Fantastic work Faye! 👏 

Some notes from the first review session:
- It would be good to document that the "parent pointers" are not stored physically, but are actually functions, in memory (and when they are set). Maybe add some high-level comment to the code and indicate it in the diagrams; they usually show "stored" physical pointers ("storage IDs" (now "slab ID" / "value ID", right?))
- Maybe document (if not already) that the inlining approach is to not inline all slabs that are "inlinable" in the sense of fitting inside of the parent, but that inlinability is determined "as a whole" for a whole value. The example was that, a small array is inlinable if it is just a single data slab, but a larger array consisting of a meta slab and some data slabs is not inlinable, even when the meta slab could fit into the parent
- Maybe document (if not already), that "slab ID" == "value ID", i.e. they share a namespace, and an ID might be used for a slab, but it might also not. Currently, a slab ID could be checked to exist
- As discussed, we might also want to update the child' parent callback when the child is inserted into the container (array and map), e.g. in the `Set` and `Insert` operations.

I'll do a proper code review after the next review session. So far the work looks great, well done! 👏

ramtinms reviewed (approved):
🚢 the logic and key parts of the changes look good to me, I did not check all the lines as we have a very good coverage by the tests. 
Besides the suggestions I put in the code, I suggest also adding this logic to the stress test or tests to check for register allocation leaks (register falls behind), especially when we continuously move from large mutable to inlineable ones. 

p.s. (cc: @janezpodhostnik ) I think we need higher-level testing for applications like the FVM environment but that's out of the scope of this repo. we might re-use the NFT tracker reporter to check the health before and after migration. and simulate some transactions to check the changes (especially re-sizing logic).

turbolent commented on array.go:
```diff
@@ -57,6 +58,15 @@ const (
+	//   value ID index head (1 byte) +
+	//   value ID index (8 bytes) +
+	//   element array head (3 bytes)
+	inlinedArrayDataSlabPrefixSize = 2 + 1 + 2 + 1 + 8 + arrayDataSlabElementHeadSize
```
Maybe we can use `unsafe.Sizeof` here on actual types?

fxamacker replied:
Not in this case because we use encoding size instead of Go type size.  We track encoding size for slab operations such as merge and split.  I added comment about this in commit 5194eee696bed5a80b9cdd42d79be430ab7a36b4.

turbolent replied:
Maybe introduce constants for the magic numbers and make the expression a bit more self-explaining. For example:
```go
tagNumberSize = 2
// ...
inlinedArrayDataSlabPrefixSize = tagNumberSize + //...
```

In addition / if possible, maybe lso define types for the individual parts, so that it is possible to use `Sizeof`, e.g.

```go
type tagNumber [2]byte

const (
	// ...
	inlinedArrayDataSlabPrefixSize = unsafe.Sizeof(tagNumber) + // ...
```

turbolent commented on array.go:
```diff
@@ -2336,22 +2664,137 @@ func NewArrayWithRootID(storage SlabStorage, rootID SlabID) (*Array, error) {
 
+// TODO: maybe optimize this
+func (a *Array) incrementIndexFrom(index uint64) {
+	for id, i := range a.mutableElementIndex {
```
`range`-ing over maps is not deterministic. In this case, the side-effects in the are independent of the order / commute to the same results, so it does not matter, but maybe add a comment here explaining that this operation is safe.

In Cadence we're using a custom linter to reject such cases (https://github.com/onflow/cadence/tree/master/tools/maprange), we might want to also use it in the atree repo?

fxamacker replied:
Yes, non-deterministic range loop is safe to use in this case and I added comment in commit f8e0992b1296afda3fd2414e82cdd24e3b6b48be.  

> In Cadence we're using a custom linter to reject such cases (https://github.com/onflow/cadence/tree/master/tools/maprange), we might want to also use it in the atree repo?

Good idea!  :+1:  I will open issue to add this custom linter to CI so we don't forget.

turbolent commented on array.go:
```diff
@@ -2336,22 +2664,137 @@ func NewArrayWithRootID(storage SlabStorage, rootID SlabID) (*Array, error) {
+
+// TODO: maybe optimize this
+func (a *Array) decrementIndexFrom(index uint64) {
+	for id, i := range a.mutableElementIndex {
```
Same as above

fxamacker replied:
See comment above.

turbolent commented on array.go:
```diff
@@ -2534,6 +2991,93 @@ func (a *Array) promoteChildAsNewRoot(childID SlabID) error {
+	inlined := a.root.Inlined()
+	inlinable := a.root.Inlinable(maxInlineSize)
+
+	if inlinable && inlined {
```
Maybe switch to a switch here, e.g.

``` go
switch {
case inlinable && inlined:
	// ...
case !inlinable && !inlined:
	// ...
// ...
```

for all 4 cases, i.e. move the last case into a explicit `case !inlinable && inlined`,
and `panic()` after the switch case to indicate an unreachable state (not all possible cases where covered)

fxamacker replied:
Good idea!  Done in commits 35849521656f4449ee79d162d41ba1543119aaa0 and dc8a567a5e743336e2f9b5fd4ecd3e61226ed639.

fxamacker commented on typeinfo.go:
```diff
@@ -32,3 +38,340 @@ type TypeInfoDecoder func(
+// Given this, composites with same type ID and same number of fields have the same fields.
+// See https://developers.flow.com/cadence/language/contract-updatability#fields
+func (ied *inlinedExtraData) addCompositeExtraData(data *MapExtraData, digests []Digest, keys []MapKey) int {
+	id := compositeTypeID{data.TypeInfo.ID(), int(data.Count)}
```
As noted by Bastian in code review meeting, new attachment types being added to Cadence need to be supported.

I think updating deduplication in atree to support attachment types may require using list of field names instead of field count.  I'm looking into it and will update this PR to support it.

fxamacker replied:
@turbolent

Done in commit 0d57f80771b790d87cad6ae125682dc028da4662.  It deduplicates composite by type ID and sorted field names.

This not only supports Cadence attachments but allows future possibilities in Cadence without modifying atree.

While at it, also refactored encoding composite type to reduce traversal and type assertion.

fxamacker commented on typeinfo.go:
```diff
@@ -19,11 +19,17 @@
 	Encode(*cbor.StreamEncoder) error
+	IsComposite() bool
+	ID() string
+	// TODO: maybe add a copy function because decoded TypeInfo can be shared by multiple slabs if not copied.
```
> is this still a relevant TODO?

Yes, good catch. The TODO was a precaution that was postponed to get PR reviews started on the other parts.

It's done in 72d3614209b6c03dd766a4b29ea1b7b98376bb69.  `TypeInfo` is copied in the commit as a precaution because `TypeInfo` can be shared among inlined slabs referring to the same type and any changes to one may affect all others if not copied.

fxamacker commented on map.go:
```diff
@@ -2264,28 +2371,285 @@ func newMapDataSlabFromDataV1(
+		firstKey: elements.firstKey(),
+	}
+
+	// TODO: does extra data needs to be copied?
```
> is this still a valid TODO?

Yes :+1:  See comment above and this is also done in the same commit https://github.com/onflow/atree/commit/72d3614209b6c03dd766a4b29ea1b7b98376bb69.

fxamacker commented:
@ramtinms @turbolent thanks for the fast PR reviews! I resolved all feedback that may require coding changes.

Ramtin, thanks for suggestions and approval!  I will open separate PR to update smoke tests.

Bastian, can you please take another look?

turbolent commented on storage.go:
```diff
@@ -34,6 +34,26 @@ const LedgerBaseStorageSlabPrefix = "$"
 
+func slabIDToValueID(sid SlabID) ValueID {
+	var id ValueID
+	copy(id[:], sid.address[:])
+	copy(id[8:], sid.index[:])
```
Better yet, consider using `unsafe.Sizeof` for the definition of `ValueID` above, in the range expression here. See for example: https://goplay.tools/snippet/EtdTvPNnIvF

turbolent commented on storable.go:
```diff
@@ -50,6 +67,14 @@ func hasPointer(storable Storable) bool {
 }
 
 const (
+	CBORTagInlinedArrayExtraData     = 247
```
We have to be careful that the two CBOR tag ranges, the one here in atree, and the one in Cadence, do not start to overlap. The unassigned range is 128-255. Atree "starts" at 255 and "grows" downward, while Cadence starts at 128 and "grows" upwards – it currently at 226. This isn't an issue yet, but maybe something we want to be proactive about.

fxamacker replied:
I added a warning comment in commit e6fa347db4ae96c0c28e8091793d09a3a562c41b.

We can add similar warning at github.com/onflow/cadence when the next Atree/Cadence integration PR is opened.  To be more proactive, we can have both repos reserve a range so they don't overlap.

turbolent commented on array.go:
```diff
@@ -78,15 +92,20 @@ type ArrayDataSlab struct {
-		root:    a,
+		Storage:             storage,
+		root:                a,
+		mutableElementIndex: make(map[ValueID]uint64),
```
Here and below: Is it possible to lazily allocate the map?

fxamacker replied:
Yes, done in commit 03acd499492eadd274209dfd14657fe04387bf5d.

turbolent commented on array.go:
```diff
@@ -482,30 +540,232 @@ func newArrayDataSlabFromDataV1(
+				len(b)))
+	}
+
+	var index [8]byte
```
Maybe define a type like `type Index [8]byte`, and use it here:
```suggestion
	var index Index
```

See also the comment below

fxamacker replied:
Done (slightly different) in commit 90be7ac4d23abbaa3128dbd076aa2338a9b12ae5.

I used `SlabIndex` instead of creating a new type.

turbolent commented on array.go:
```diff
@@ -482,30 +540,232 @@ func newArrayDataSlabFromDataV1(
+
+	extraDataIndex := inlinedTypeInfo.addArrayExtraData(a.extraData)
+
+	if extraDataIndex > 255 {
+		return NewEncodingError(
+			fmt.Errorf("failed to encode inlined array data slab: extra data index %d exceeds limit 255", extraDataIndex))
```
Maybe define a constant for this magic number here

```suggestion
	if extraDataIndex > maxArrayExtraDataIndex {
		return NewEncodingError(
			fmt.Errorf("failed to encode inlined array data slab: extra data index %d exceeds limit %s", extraDataIndex, maxArrayExtraDataIndex))
```

turbolent commented on array.go:
```diff
@@ -482,30 +540,232 @@ func newArrayDataSlabFromDataV1(
+	inlinedTypes := newInlinedExtraData()
+
+	// TODO: maybe use a buffer pool
+	var elementBuf bytes.Buffer
```
Would it be possible to avoid using/allocating a buffer here, and just write and "patch up" the written  data afterwards? For example, by writing out-of-order and then re-ordering the written data

fxamacker replied:
No, unfortunately we can't (easily) just write and "patch up" the written data afterwards.  For example, `Encoder` uses a writer under the hood and writer can't be patched up so we would need to replace writer with something.

Given this, it is easier to use a buffer pool to reduce allocations as noted in the TODO, so I used buffer pool in commit 836eb70922e82d00723b90ebf022275be5fb74a9.

turbolent commented on array.go:
```diff
@@ -2336,22 +2670,145 @@ func NewArrayWithRootID(storage SlabStorage, rootID SlabID) (*Array, error) {
+
+		// Verify overwritten storable has identical value ID.
+
+		switch x := existingValueStorable.(type) {
```
Maybe just "rebind" the variable to increase readability:
```suggestion
		switch existingValueStorable := existingValueStorable.(type) {
```
and use  `existingValueStorable` instead of `x` in the cases

turbolent commented on array.go:
```diff
@@ -2382,6 +2839,15 @@ func (a *Array) Set(index uint64, value Value) (Storable, error) {
 		}
 	}
 
+	err = a.notifyParentIfNeeded()
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up notification callback in child value so
+	// when child value is modified parent a is notified.
+	a.setCallbackWithChild(index, value, maxInlineArrayElementSize)
```
I'm not quite sure about when notification is necessary, but should it happen before the callback update? Should it happen afterwards instead? or maybe also afterwards in addition to before?

fxamacker replied:
I added more comments in a75e3886e2987a452487a5ed02559c284a4b096c.  

- `notifyParentIfNeeded` is called when receiver (as a child) is modified.  For example, in `Array.Set()` and `Array.Remove()`, array (as a child) is modified, while in `Array.Get()` array (as a child) isn't modified.

- `setCallbackWithChild` is called when receiver (as a parent) needs to communicate with child value.  For example, in `Array.Get()`, array (as a parent) wants to communicate with child value. while in `Array.Remove()`, array (as a parent) doesn't need to communicate with removed value.

turbolent replied:
Perfect, thank you!

turbolent commented on map.go:
```diff
@@ -3888,8 +4606,99 @@ func NewMapWithRootID(storage SlabStorage, rootID SlabID, digestBuilder Digester
+
+		// Verify overwritten storable has identical value ID.
+
+		switch x := existingValueStorable.(type) {
```
```suggestion
		switch existingValueStorable := existingValueStorable.(type) {
```

turbolent commented on map.go:
```diff
@@ -2264,91 +2405,518 @@ func newMapDataSlabFromDataV1(
+			fmt.Errorf("failed to encode standalone map data slab as inlined"))
+	}
+
+	if hkeys, keys, values, ok := m.canBeEncodedAsComposite(); ok {
```
Small naming nit: Before the review session, it wasn't clear to me why atree's ordered map encoding handled composites (a Cadence concept). As far as I remember the explanation in the session, the idea was that composites with same fields and types could be encoded in a more compact form (deduplicated types and field names, right?). 

Maybe name it "Compact" instead of "Composite"? The latter is a bit confusing from a Cadence perspective, because there, (Cadence) composites are always stored as ordered maps, but then here atree decides on a particular representation for the ordered map.

fxamacker replied:
Good point!  I renamed "Composite" to "CompactMap" in commit 4b3c26cb860e418238e7791842ec746248a46f79.

turbolent commented on map.go:
```diff
@@ -2264,91 +2405,518 @@ func newMapDataSlabFromDataV1(
+
+	extraDataIndex := inlinedTypeInfo.addMapExtraData(m.extraData)
+
+	if extraDataIndex > 255 {
+		return NewEncodingError(fmt.Errorf("extra data index %d exceeds limit 255", extraDataIndex))
```
```suggestion
	if extraDataIndex > maxMapExtraDataIndex {
		return NewEncodingError(fmt.Errorf("extra data index %d exceeds limit maxMapExtraDataIndex", extraDataIndex))
```

turbolent commented on map.go:
```diff
@@ -2264,91 +2405,518 @@ func newMapDataSlabFromDataV1(
 
-	if m.extraData != nil {
-		h.setRoot()
+	if extraDataIndex > 255 {
+		// This should never happen because of slab size.
+		return NewEncodingError(fmt.Errorf("extra data index %d exceeds limit 255", extraDataIndex))
```
```suggestion
	if extraDataIndex > maxMapExtraDataIndex {
		// This should never happen because of slab size.
		return NewEncodingError(fmt.Errorf("extra data index %d exceeds limit 255", maxMapExtraDataIndex))
```

turbolent reviewed (commented):
Nice! 

Mainly wondering about the parent callback updates and notifications. Maybe we can clarify async, but also happy to have a quick sync about it and talk it through

fxamacker commented:
@turbolent  I found and resolved edge case found while adding comments about how inlining works on Sunday.

While at it, I also refactored, added more checks, and added no-op to parentUpdater we discussed last week.  Thanks again for good discussions last week.  :+1:

turbolent commented:
Fantastic work Faye! 👏 

Here are the cleaned up notes I took from the last review session last Thursday. They might not be up-to-date anymore, as you might already have addressed some points since then, so please feel free to disregard if that is the case.

- Maybe add some general/conceptual comments/explanations of how the inlining works, and what operations/cases need to be supported.
  - For example, children need to notify their parent on any mutation.
  - For example, for maps, there are two operations that mutate, set and remove
  - After a mutation, a receiver/container needs to:
    - Notify its parent (if any), because the receiver/container might be itself a child of another container (with `notifyParentIfNeeded`)
    - Update the child's parent callback, so that the inserted/removed child can notify the receiver/container (with `setCallbackwithChild`) when that child gets mutated

- In the array/map remove operations, only a storable is returned. Similar, the set operations returns a potentially existing existing storable for the given key.

  However, there might be values for the storable which have parent callbacks. Those callbacks should theoretically be unset. Loading the value for the storable does not make much sense, as current implementations will allocate another, new value. 

  I don't quite remember the plan for resolving this. Was the plan to leave the parent callback happening, detect the "invalid" child callback to the "previous" parent, and reject it?

- Sometimes "value" is used in a variable name, even though it refers to a just a storable. For example, there was a case for `extistingValue`, which was only a `Storable`. Maybe check naming of such variables and update accordingly to avoid confusion.

- We walked through the example case of inserting a child into a parent, mutating it, removing the child from that parent, mutating it, and then inserting the child into another container, and mutating it.

  - It would be good to have an integration test for this in both atree and Cadence.

  - We noticed that a mutation after the removal (or set operation), atree needs to uninline the child

  - A callback of a child to a previous parent was a fatal error. However, this is likely not an invalid case:
    The child value is mutated, so notifies its parent (parent callback was not unset, but also could not, see above). 
    Instead of a fatal error, this callback should be a "no-op". Further, when the parent detects such a callback from a previous child, it should now use this opportunity to unset the parent callback from the child.

- We talked through the "singleton value" case: `StoredValue` might return same, existing value, and might not return a new value. It might be good to not make any assumptions about the nature of the returned value (e.g. that an implementation is required to always return a new value), or at least document the assumptions properly. 

  This prevents regressions in the future, where such assumptions might be broken by the user of the library, i.e. Cadence.

- On the Cadence side, there is the implicit invariant that all values stored in an account always reference values in the same account, and never across accounts.
  We should add defensive assertions for this invariant, i.e. check check that the container and child addresses match.

- Finally, we discussed the encoding, in particular, the deduplication of type information.
  
  When types are dedupliated, there seeds are normalized.

  In atree comments/docstring, in Cadence comments/docstrings, and in the Cadence user documentation, we should emphasize that iteration order is undefined (but deterministic of course), and no assumptions should be made during execution or across transactions. For example, just because iteration over the keys of a dictionary has a particular order in one transaction does not mean that it will be the same in the next transaction.

  The deduplication optimization will change the current behavior, which some users might rely on (but should (have) not).

I will have another look over the changes of the PR today and will open issues for the TODOs on the Cadence side mentioned above.

turbolent commented on utils_test.go:
```diff
@@ -100,13 +113,50 @@ func (i testTypeInfo) Equal(other TypeInfo) bool {
 func typeInfoComparator(a, b TypeInfo) bool {
-	x, ok := a.(testTypeInfo)
-	if !ok {
+	switch x := a.(type) {
+	case testTypeInfo:
+		return x.Equal(b)
+
+	case testCompositeTypeInfo:
+		return x.Equal(b)
```
```suggestion
	switch a := a.(type) {
	case testTypeInfo:
		return a.Equal(b)

	case testCompositeTypeInfo:
		return a.Equal(b)
```

turbolent commented on array.go:
```diff
@@ -2857,7 +2881,18 @@ func (a *Array) notifyParentIfNeeded() error {
+		return err
+	}
+	if !found {
+		a.parentUpdater = nil
```
👍

turbolent commented on array.go:
```diff
@@ -2341,21 +2776,236 @@ func NewArrayWithRootID(storage SlabStorage, rootID SlabID) (*Array, error) {
+	// storage and any future changes to it would have been lost.
+	switch s := existingStorable.(type) {
+	case ArraySlab:
+		err = s.Uninline(a.Storage)
```
Should this be only done if it is inlined? `Uninline` seems to error when that is the case

fxamacker replied:
> Should this be only done if it is inlined?

Yes, it is only done for inlined slabs.  

Only inlined array slab falls into `case ArraySlab`.  Not-inlined slab (stored as `SlabIDStorable`) falls into `case SlabIDStorable`.

turbolent replied:
Ah, that makes sense. Could you please add this as a comment? I didn't quite see that at first glance

turbolent reviewed (commented):
Great work addressing all the feedback! 👏 
The changes are looking good to me 👍

turbolent commented:
@fxamacker Would be good to sync on remaining work, see the (sorry, long) comment above. If there isn't really anything left to be addressed, this should be good to go 🎉 

Additional tests and adjustments (if needed) can be done in follow-up PRs, this one is already very large

turbolent commented:
I had a look at adding the container/child address matching checks to Cadence: Just before performing a mutating operation on a container (array/map), we already transfer the child, so checking the invariant does not really help much: if we insert the assertion we can also check the transfer.

On the atree side, it's not easy to check this invariant: children are values at first, and getting the storable (which is needed to get the potential storage ID) is delayed until the very "end", e.g. deep down when the container/parent is not available anymore.

Maybe there isn't much we can do here

fxamacker commented:
@turbolent  Thanks for detailed review and meeting notes!

I think everything discussed so far have been addressed in this PR.  :tada:

- More comments were added about how atree inlining works in commit a75e3886e2987a452487a5ed02559c284a4b096c

- Commit 63ea7a74dd58db15fc871c40f4c3c47772dc8c81 handles values with outdated parent callbacks (value was removed or overwritten in parent container):
  - the callback is no-op when callback detects value is no longer a child in its old parent.  
  - value unsets outdated parent callback 

- Commit 8d02bbcd024cf77a14b6feb31ca7a4aadec330c7 renamed `existingValue`.

- Commit 042eb686fc19f48fe3e78a69b3cbe6de979a4001 unlined removed or overwritten child (if it is inlined)

- I double-checked that atree doesn't make assumption about value returned from `StoredValue()`.  

- About "invariant that all values stored in an account always reference values in the same account"
  - it isn't easy to check in code as you mentioned.  However, we check this in tests, for example, `atree.CheckStorageHealth()` checks that all child or referenced slabs have the same address and atree validation functions also checks inlined slabs in memory have the same address as its parent.  

- I opened these issues to add more integration tests:
  - https://github.com/onflow/atree/issues/346 
  - https://github.com/onflow/cadence/issues/2837 

BTW, can you please take a look at PR #345?  That is the 2nd of 2 atree inlining PRs.

turbolent reviewed (approved):
Fantastic work Faye! 👏 

My review focused mainly on the changes to array and ordered map, I didn't review the tests, but they seem very thorough (18 kLOC wow!) 👏
